src/server.rs

```diff
@@ -0,0 +1,395 @@
+use std::borrow::Cow;
+use std::io;
+
+use data_encoding::base64;
+use rand::distributions::IndependentSample;
+use rand::distributions::range::Range;
+use rand::os::OsRng;
+use rand::Rng;
+use ring::digest::Digest;
+
+use error::{Error, Kind, Field};
+use utils::find_proofs;
+use ::{NONCE_LENGTH, SHA256_LEN};
+
+/// Responds to client authentication challenges.
+/// The entrypoint for the SCRAM server side implementation
+pub struct ScramServer<P: AuthenticationProvider> {
+    /// The [authentication provider](trait.AuthenticationProvider.html) that will find passwords
+    /// and check authorization
+    provider: P
+
+}
+
+/// Contains information about stored passwords.  In particular, it stores the password that
+/// has been salted and hashed, the salt that was used, and the number of iterations of the hashing
+/// algorithm
+pub struct PasswordInfo {
+    hashed_password: Vec<u8>,
+    salt: Vec<u8>,
```

Comment:
I think that this should store bytes slices instead of vectors. I have two arguments for this change:

* The ownership of these values is often held by other data structures. Taking a slice avoids copies (like in [tests/client_server.rs line 27](https://github.com/tomprogrammer/scram/pull/3/files#diff-63485cab5eb29b93e2de36ed0e68866cR27)).
* Salt and hashed password can be stored together and sliced individually. Short example:
```
const SALT_LENGHT: usize = 20;
let (hashed_password, iterations) = password_db.get(&username).unwrap();
let password_info = PasswordInfo::new(&hashed_password[SALT_LENGTH..], 
                                      iterations,
                                      &hashed_password[..SALT_LENGTH]);
```